runtime/src/swam/runtime/internals/instance/MemoryInstance.scala

```diff
@@ -40,47 +40,91 @@ private[runtime] class MemoryInstance[F[_]](min: Int, max: Option[Int], onHeap:
         ByteBuffer.allocate(size)
       else
         ByteBuffer.allocateDirect(size)
+
+    if (tracer != null)tracer.traceEvent("msize", size)
     buffer.order(ByteOrder.LITTLE_ENDIAN)
     buffer
   }
 
   def size = buffer.capacity
 
-  def writeByte(idx: Int, v: Byte) =
+  def writeByte(idx: Int, v: Byte) = {
+    if (tracer != null)tracer.traceEvent("msize", size)
     F.delay(buffer.put(idx, v))
+  }
 
-  def readByte(idx: Int) =
-    F.delay(buffer.get(idx))
+  def readByte(idx: Int) = {
+    F.delay({
```

Comment:
style: you can drop the parentheses around the braces, which makes it look nicer:
```scala
F.delay {
  ...
}
```